batch-index-files/main.py

```diff
@@ -8,55 +8,81 @@
+        print(f"retrying {len(files_to_retry)} files")
+        files_to_retry = process_files_concurrently(boto_s3, bucket, delete, files_to_retry)
+        print(f"remaining files after second pass {len(files_to_retry)}")
+    if len(files_to_retry) == 0:
+        bio_idx_host = 'http://ec2-18-215-38-136.compute-1.amazonaws.com:5001'
+        requests.post(f"{bio_idx_host}/bgcompress/mark-completed/{index}?prefix={path}")
+
+
+def process_files_concurrently(boto_s3, bucket, delete, files):
+    files_to_retry = []
+    print_lock = threading.Lock()
     with concurrent.futures.ThreadPoolExecutor(max_workers=60) as executor:
         futures = []
-        for file in s3_objects:
-            futures.append(executor.submit(bg_compress_and_index_file, bucket, file['Key'], boto_s3, delete))
+        for file in files:
+            futures.append(executor.submit(bg_compress_and_index_file, bucket, file, boto_s3, delete,
+                                           files_to_retry, print_lock))
         # wait for all futures to complete
         for future in concurrent.futures.as_completed(futures):
             future.result()
+    return files_to_retry
 
 
 def get_access_keys():
     client = Session().client('secretsmanager')
     return json.loads(client.get_secret_value(SecretId='bgzip-credentials')['SecretString'])
 
 
-def bg_compress_and_index_file(bucket_name, file, boto_s3, delete_json_file):
-    error_messaage = None
+def bg_compress_and_index_file(bucket_name, file, boto_s3, delete_json_file, files_to_retry, print_lock):
+    error_message = None
     results = boto_s3.list_objects(Bucket=bucket_name, Prefix=file + ".gz")
     if results.get('Contents', None) and len(results.get('Contents')) == 2:
-        print(f"Compressed index file already exists: {file}")
+        with print_lock:
+            print(f"Compressed index file already exists: {file}")
+        if delete_json_file:
+            boto_s3.delete_object(Bucket=bucket_name, Key=file)
         return
-    print(f"starting {file}")
+    with print_lock:
+        print(f"starting {file}")
     command = ['bgzip', '-i', f"s3://{bucket_name}/{file}"]
     try:
-        subprocess.run(command, check=True, timeout=120)
+        subprocess.run(command, check=True, timeout=120, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
         if delete_json_file:
             boto_s3.delete_object(Bucket=bucket_name, Key=file)
     except subprocess.CalledProcessError as e:
-        error_messaage = f"Error: Command exited with non-zero status: {e.returncode}, {file}"
+        error_message = f"Error: Command exited with non-zero status: {e.returncode}, {file}"
     except subprocess.TimeoutExpired as e:
-        error_messaage = f"Error: Command timed out after {e.timeout} seconds"
+        error_message = f"Error: Command timed out after {e.timeout} seconds"
 
-    if error_messaage:
-        print(error_messaage)
+    if error_message:
+        with print_lock:
```

Comment:
I was previously printing without a lock and I noticed that I was occasionally getting unprintable characters on the console.  Making sure one thread at a time calls print seems to have fixed things.